### PR TITLE
Fix build being called before element is ready to build.

### LINKED ISSUE #4397: buildCallback invoked when hasNextNodeInDocumentOrder is false
screenshot from 2016-08-05 15 23 27

It's a bit hard to reproduce but as you can see in the screenshot, the `buildCallback` of `<amp-experiment>` is invoked when the `children=[]`. `hasNextNodeInDocumentOrder` also returns `false`. A snapshot of the DOM is printed in the console. We can see at that point, `<amp-experiment>` is not fully ready yet.

To reproduce this locally, please patch the diff: #4399
You can set the same break point as me, and refresh the page a couple times. Normally you see it less than 10 refreshes.

The call stack (on the top right corner of the screenshot) shows the `build` was triggered via 

```
    this.visibilityStateMachine_.setState(
      this.viewer_.getVisibilityState()   // = visible
    );
```

@dvoytenko @mkhatib do you have any idea?

+@rudygalfi  this is what we met in the ABE amp-experiment article.

### PR DESCRIPTION
**Pull request by mkhatib**
Fixes #4397

### CODE REVIEW

**Comment by mkhatib** [comment]
I was able to test with and without this fix on @lannka patch #4399 and was unable to reproduce the issue with this fix in place.

File: src/service/resources-impl.js [hunk 1]
```diff
@@ -932,7 +941,7 @@ export class Resources {
     for (let i = 0; i < this.resources_.length; i++) {
       const r = this.resources_[i];
```
Comment (lannka) [inline src/service/resources-impl.js]: could you rename `r` to `resource`? It would have made code search easier :-).

**Comment by lannka** [comment]
LGTM defer to @dvoytenko for the approval

File: src/service/resources-impl.js [hunk 2]
```diff
@@ -336,20 +336,30 @@ export class Resources {
...
+  }
 
+  /**
+   * Builds the element if ready to be built, otherwise adds it to pending resources.
+   * @private
+   */
+  buildOrScheduleBuildForElement_(element) {
+    const resource = this.getResourceForElement(element);
     if (this.isRuntimeOn_) {
+      if (element.isBuilt()) {
```
Comment (dvoytenko) [inline src/service/resources-impl.js]: Move to the very top.

Comment (dvoytenko) [inline src/service/resources-impl.js]: One thing, however. You are loosing `schedulePass` on `add` in this case. Intended? Is there a reason why you wanted to change the old behavior?

File: src/service/resources-impl.js [hunk 3]
```diff
@@ -1001,7 +1010,7 @@ export class Resources {
     // Phase 3: Trigger "viewport enter/exit" events.
     for (let i = 0; i < this.resources_.length; i++) {
       const r = this.resources_[i];
-      if (r.hasOwner()) {
+      if (r.getState() == ResourceState.NOT_BUILT || r.hasOwner()) {
```
Comment (dvoytenko) [inline src/service/resources-impl.js]: Can't make a comment below, but I believe we also need to check the build status in the following code below:

`if (r.getState() != ResourceState.READY_FOR_LAYOUT || r.hasOwner()) {`

Comment (mkhatib) [inline src/service/resources-impl.js]: I feel that would be repetitive. Adding it anyway.

File: src/service/resources-impl.js (hunk 1 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Filed an issue to track some refactoring here: #4450 - don't wanna get this PR too complicated - wanna check it in for the upcoming canary.

File: src/service/resources-impl.js (hunk 2 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: You meant on `upgraded` not `add`, right? 

I was thinking the `schedulePass` inside this new function would take care of that, but I guess it's not the same, putting it back in upgraded call.

Is there harm from always scheduling a pass at the end of `buildOrScheduleBuildForElement_` but inside the `if(isRuntimeOn)` check instead?

File: src/service/resources-impl.js [hunk 4]
```diff
@@ -336,7 +336,16 @@ export class Resources {
...
     this.resources_.push(resource);
+    this.buildOrScheduleBuildForElement_(element);
+    dev().fine(TAG_, 'element added:', resource.debugid);
+  }
 
+  /**
+   * Builds the element if ready to be built, otherwise adds it to pending resources.
+   * @private
+   */
+  buildOrScheduleBuildForElement_(element) {
```
Comment (dvoytenko) [inline src/service/resources-impl.js]: Let's switch and pass `resource` to this method. And please add jsdoc for it.

**Comment by dvoytenko** [comment]
LGTM pending tests.

File: src/service/resources-impl.js (hunk 4 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Done

**Pull request merged**